### PR TITLE
Extension of PR#105

### PR DESCRIPTION
**Pull request by p-wein**
## Short Description
This PR is based on PR #105 and fixes some small mistakes, e.g. block_size_y and block_size_z were sometimes based on img_rows. Extends the Genesis transforms to accept absolute or relative patch sizes and all Inpainting transforms to accept custom border distances (was set to 3 in #105). Adds separate patch sizes for in and outpainting in  RandomInOrOutpainting.
Since this is one of my first PRs, I am not sure if everything is done properly. So please double check everything and feel free to comment if anything is unclear.
Cheers

### PR Implementer
This is a small checklist for the implementation details of this PR.
If you submit a PR, please look at these points (don't worry about the `RisingTeam`
and `Reviewer` workflows, the only purpose of those is to have a compact view of
the steps). I did not implement unit tests but did some short test for all changed transforms, (ran with default and sometimes additional settings on images of size (32,192,192) for 100 iterations).

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/PhoenixDL/rising/blob/master/CONTRIBUTING.md).

- [x] Implementation
- [x] Docstrings & Typing
- [x] Check `__all__` sections and `__init__`
- [ ] Unittests (look at the line coverage of your tests, the goal is 100%!)
- [ ] Update notebooks & documentation if necessary
- [ ] Pass all tests
- [ ] Add the checksum of the last implementation commit to the Changelog

### RisingTeam
<details>
  <summary>RisingTeam workflow</summary>

  - [ ] Add pull request to project (optionally delete corresponding project note)
  - [ ] Assign correct label (if you don't have permission to do this, someone will do it for you. 
      Please make sure to communicate the current status of the pr.)
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)
</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>
  
  - [ ] Do all tests pass? (Unittests, NotebookTests, Documentation)
  - [ ] Does the implementation follow `rising` design conventions?
  - [ ] Are the tests useful? (!!!) Are additional tests needed? 
        Can you think of critical points which should be covered in an additional test?
  - [ ] Optional: Check coverage locally / Check tests locally if GPU is necessary to execute
</details>

### CODE REVIEW

File: rising/transforms/functional/painting.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+def local_pixel_shuffle(data: torch.Tensor, n: int = -1, block_size: tuple = (0, 0, 0), rel_block_size: float = 0.1) -> torch.Tensor:
+
+    batch_size, channels, img_rows, img_cols, img_deps = data.size()
+
+    if n < 0:
+        n = int(1000 * channels)  # changes ~ 12.5% of voxels
+    for b in range(batch_size):
+        for _ in range(n):
+            c = torch.randint(0, channels - 1, (1,))
```
Comment (p-wein) [inline rising/transforms/functional/painting.py:14]:
```suggestion
            c = torch.randint(0, max(1, channels - 1), (1,))
```
If single channel/modality input (e.g. T2 only) was used this errored, since max=min=0 and torch.randint requires max> min.

File: rising/transforms/spatial.py [hunk 2]
```diff
@@ -39,8 +38,30 @@ def __init__(self,
...
-                         property_names=('dims',), **kwargs)
+        super().__init__(grad=grad, **kwargs)
+        self.keys = keys
+        self.prob = prob
+        if not isinstance(dims, DiscreteParameter):
+            if len(dims) > 2:
+                dims = list(combinations(dims, 2))
+            else:
+                dims = (dims,)
+            dims = DiscreteParameter(dims)
```
Comment (p-wein) [inline rising/transforms/spatial.py:49]: Missing correct handling for type(dims)==int. Will throw error on line 45

File: rising/utils/torchinterp1d.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
+
+import torch
+import contextlib
+SEARCHSORTED_AVAILABLE = True
+try:
+    from torchsearchsorted import searchsorted
```
Comment (justusschock) [inline rising/utils/torchinterp1d.py:38]: can you add a python implementation of the search sorted functionality?

Comment (p-wein) [reply rising/utils/torchinterp1d.py:38]: This is a problem from #105 (see https://github.com/PhoenixDL/rising/pull/105#discussion_r435235907) by @weningerleon so I am not sure what is done there exactly, but to me it seems like this was only necessary for pytorch < v1.6, since 1.6 added https://pytorch.org/docs/stable/generated/torch.searchsorted.html?highlight=searchsorted#torch.searchsorted and https://pytorch.org/docs/stable/generated/torch.bucketize.html#torch.bucketize.

File: rising/utils/torchinterp1d.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+        # moving everything to the desired device in case it was not there
+        # already (not handling the case things do not fit entirely, user will
+        # do it if required.)
+        for name in v:
+            v[name] = v[name].to(device)
+
+        # calling searchsorted on the x values.
+        ind = ynew.long()
+        searchsorted(v['x'].contiguous(), v['xnew'].contiguous(), ind)
```
Comment (p-wein) [inline rising/utils/torchinterp1d.py:143]:
```suggestion
        torch.searchsorted(v['x'].contiguous(), v['xnew'].contiguous(), ind)
```

File: rising/utils/torchinterp1d.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
+
+import torch
+import contextlib
+SEARCHSORTED_AVAILABLE = True
+try:
+    from torchsearchsorted import searchsorted
+except ImportError:
+    SEARCHSORTED_AVAILABLE = False
```
Comment (p-wein) [inline rising/utils/torchinterp1d.py:40]:
```suggestion

```

File: rising/utils/torchinterp1d.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+            Tensor for the output. If None: allocated automatically.
+
+        """
+        # checking availability of the searchsorted pytorch module
+        if not SEARCHSORTED_AVAILABLE:
+            raise Exception(
+                'The interp1d function depends on the '
+                'torchsearchsorted module, which is not available.\n'
+                'You must get it at ',
+                'https://github.com/aliutkus/torchsearchsorted \n')
```
Comment (p-wein) [inline rising/utils/torchinterp1d.py:79]:
```suggestion
```

File: rising/utils/torchinterp1d.py (hunk 4 above)
Comment (p-wein) [reply rising/utils/torchinterp1d.py:143]: To be honest I have no clue how torch.searchsorted behaves compared to torchsearchsorted.searchsorted . Please check this yourselves as this is just an "educated" guess.

Comment (weningerleon) [reply rising/utils/torchinterp1d.py:143]: I'll have a look at it...

Comment (weningerleon) [reply rising/utils/torchinterp1d.py:143]: Yep it works, I updated the torchinterp1d file, no need for the old dependency any more thanks to pyTorch 1.6
I updated the other pull request. I guess it would be best if we just have one pull request, instead of two with the same changes?

Comment (p-wein) [reply rising/utils/torchinterp1d.py:143]: Yes, I agree. 
I think it would be best to have a single PR. I only created a new one, as I could not commit to yours (sorry, I am still a bit confused about github and PRs). I propose that I rebase my additions on your PR again and change the default parameters as follows and according to your proposal https://github.com/PhoenixDL/rising/pull/105#discussion_r510049246: I would like to set all absolute patch sizes to tuples of 0 and only use relative patch_sizes if the absolute patch sizes are tuples of 0 (or each element <= 0). This would result in relative patch sizes being the default and users can overwrite them with absolute patch sizes without having to change relative patch sizes.
I think you need to allow me to commit to your PR though. Let me know what you think :)
